### DISCUSSION TITLE
Improving Multilingual Support and Image Processing with Tika Engine in Open WebUI: Challenges and Recommendations

### DISCUSSION THREAD

**Discussion by ahm631** [open-webui/open-webui#7297] [General]
I would like to express my sincere appreciation for your outstanding efforts in developing Open WebUI. I have been utilizing Tika as a file-reading engine, and it performs excellently with English text. However, I have encountered some challenges when working with other languages, such as Arabic.

1. OCR Challenges with Arabic Language

After installing the tesseract-ocr-ara package on Tika, I did not observe any effect within the Open WebUI interface. Through testing with Postman, I discovered that it is necessary to specify the OCR language of the image by passing it via the X-Tika-OCRLanguage parameter. Currently, Open WebUI does not provide an option to set this language, and it seems that the system defaults to sending the primary language as English (eng).

I attempted to modify Tika by replacing tesseract-ocr-eng with tesseract-ocr-ara, which resulted in successful operation within Open WebUI. However, when the file contains English text, the system is unable to read it. Additionally, there is an issue when multiple languages are present in the file.

Therefore, I suggest implementing a feature in Open WebUI that allows users to specify the languages present in files or images before sending them to the Tika server, ensuring more accurate results.

2. Processing Standalone Images

When uploading an image directly to Open WebUI and querying the language model about it, the system is unable to read the content. In contrast, if the image is embedded within a PDF file, it is processed successfully with OCR, particularly for English text.

I recommend enhancing the support for processing standalone images in Open WebUI, enabling the system to apply OCR directly to images without the need to embed them in PDF files.

Thank you once again for your dedication, and I look forward to future improvements that will enhance multilingual support and image processing capabilities in Open WebUI.

**Comment by meepuru** [comment]
Since Tika uses http api, there is an easy way to modify requests of open-webui using web proxy server like nginx:
```nginx
server {
    listen 127.0.0.1:9999;

    location / {
        # Modify header to language(s) you want to specify.
        proxy_set_header X-Tika-OCRLanguage "chi_sim+eng";
        # Proxy pass to your Tika instance.
        proxy_pass http://127.0.0.1:9998;
    }
}
```
Then just point url of Tika to `http://127.0.0.1:9999` in configuration of open-webui.
Hope this workaround can help.

**Reply by rakehell1986** [reply]
> Since Tika uses http api, there is an easy way to modify requests of open-webui using web proxy server like nginx:
> 
> ```nginx
> server {
>     listen 127.0.0.1:9999;
> 
>     location / {
>         # Modify header to language(s) you want to specify.
>         proxy_set_header X-Tika-OCRLanguage "chi_sim+eng";
>         # Proxy pass to your Tika instance.
>         proxy_pass http://127.0.0.1:9998;
>     }
> }
> ```
> 
> Then just point url of Tika to `http://127.0.0.1:9999` in configuration of open-webui. Hope this workaround can help.

I got 502 issue.....


server {
    listen 9097;
    server_name 127.0.0.1;

    location / {
        # Modify header to language(s) you want to specify
        proxy_set_header X-Tika-OCRLanguage "chi_sim+eng";
        # Proxy pass to your tika instance.
        proxy_pass http://localhost:9997;
    }


}

**Reply by meepuru** [reply]
Hmmm, it is a bit out of topic but of course I will try to help you with that.

Using a `127.0.0.1` as server name is a little confusing here, but that's not the point.
First, make sure your tika instance is indeed listening on port `9997` instead of the default `9998`. If so, avoid using `localhost` as backend address, use `127.0.0.1` instead. The reason is `localhost` might resolve as IPv6 loop back address that tika instance not listening on.

If this doesn't work, please provide more complete information so I can provide further assistance.

**Reply by rakehell1986** [reply]
> Hmmm, it is a bit out of topic but of course I will try to help you with that.
> 
> Using a `127.0.0.1` as server name is a little confusing here, but that's not the point. First, make sure your tika instance is indeed listening on port `9997` instead of the default `9998`. If so, avoid using `localhost` as backend address, use `127.0.0.1` instead. The reason is `localhost` might resolve as IPv6 loop back address that tika instance not listening on.
> 
> If this doesn't work, please provide more complete information so I can provide further assistance.

I provide my confi as following 

Docker run nginx, tika, openwebui

```
1fa6c3435f55   nginx:latest                                 "/docker-entrypoint.…"   12 hours ago   Up 12 hours             0.0.0.0:80->80/tcp, :::80->80/tcp, **0.0.0.0:9097->9097**/tcp, :::9097->9097/tcp   nginx-nginx-1
8e3cc481d8d9   apache/tika:latest-full                      "/bin/sh -c 'exec ja…"   15 hours ago   Up 15 hours             0.0.0.0:**9997**->9998/tcp, :::9997->9998/tcp                                      open-webui-tika-1
f0710f52160a   ghcr.io/open-webui/open-webui:main           "bash start.sh"          17 hours ago   Up 15 hours (healthy)   0.0.0.0:8020->8080/tcp, :::8020->8080/tcp                                      open-webui-openwebui-1
```

config in openwebui for doc

http://host.docker.internal:9097 also tried http://nginx:9097, http://127.0.0.1:9097

config in nginx 


```
[admin@iZuf61gvxer7wa7284w5u9Z nginx]$ cat ./conf/conf.d/default.conf 
server {
    listen 9097;
    server_name 127.0.0.1;

    location /  {
        # Modify header to language(s) you want to specify
        proxy_set_header X-Tika-OCRLanguage "chi_sim+eng";
        # Proxy pass to your tika instance.
        proxy_pass http://127.0.0.1:9997;
    }


}

server {
    listen       80;
    listen  [::]:80;
    server_name  localhost;
```


and then test with curl on local machine : 

```
[admin@iZuf61gvxer7wa7284w5u9Z nginx]$ curl -X GET http://localhost:9997/tika
This is Tika Server (Apache Tika 3.1.0). Please PUT
[admin@iZuf61gvxer7wa7284w5u9Z nginx]$ curl -X GET http://localhost:9097/
<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx/1.27.4</center>
</body>
</html>
[admin@iZuf61gvxer7wa7284w5u9Z nginx]$ 
```

**Comment by rakehell1986** [comment]
resolve the configuration issue.  give the 127.0.0.1 for listening , and give the host IP for proxy_pass instead of 127.0.0.1。 
the ip given in webui page is host.docker.internal。

but proxy_set_header X-Tika-OCRLanguage "chi_sim+eng"; seems wont solve the OCR scann issue. 
my scanned pdf in chinese is not correctly recognized.

**Reply by rakehell1986** [reply]
go into docker container， docker --user root exec -it connainer_name /bin/bash
apt-get update
apt-get install tessera-ocr-chi-sim

**Comment by MrDick-11** [comment]
So the only option is using nginx?
Will there be support for other languages configured in open webui for Tika?
Is there built in support for other languages in open webui without Tika?
